src/Middleware/OutputCaching/src/RecyclableSequenceBuilder.cs

```diff
@@ -0,0 +1,127 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.AspNetCore.OutputCaching;
+
+// allows capture of written payloads into a ReadOnlySequence<byte> based on RecyclableReadOnlySequenceSegment
+internal sealed class RecyclableSequenceBuilder : IDisposable
+{
+    RecyclableReadOnlySequenceSegment? _firstSegment, _currentSegment;
+    private int _currentSegmentIndex;
+    private readonly int _segmentSize;
+    private bool _closed;
+
+    public long Length { get; private set; }
+
+    internal RecyclableSequenceBuilder(int segmentSize)
+    {
+        if (segmentSize <= 0)
+        {
+            throw new ArgumentOutOfRangeException(nameof(segmentSize), segmentSize, $"{nameof(segmentSize)} must be greater than 0.");
```

Comment:
Use the new throw-helpers (ThrowIfNegative)?

Reply:
Couldn't see a ThrowIfNegativeOrZero - I guess I could subtract one...

Reply:
https://learn.microsoft.com/en-us/dotnet/api/system.argumentoutofrangeexception.throwiflessthanorequal?view=net-8.0

Reply:
heh, there literally is a `ThrowIfNegativeOrZero` - must have missed that (or a multi-TFM thing)